src/arithmetic/algebraic_expression/algebraic_expression_eval.c

```diff
@@ -0,0 +1,236 @@
+/*
+* Copyright 2018-2019 Redis Labs Ltd. and Contributors
+*
+* This file is available under the Redis Labs Source Available License Agreement
+*/
+
+#include "utils.h"
+#include "../algebraic_expression.h"
+
+// Forward declarations
+GrB_Matrix _AlgebraicExpression_Eval(const AlgebraicExpression *exp, GrB_Matrix res);
+
+static GrB_Matrix _Eval_Transpose
+(
+	const AlgebraicExpression *exp,
+	GrB_Matrix res
+) {
+	assert(exp && AlgebraicExpression_ChildCount(exp) == 1);
+
+	AlgebraicExpression *child = exp->operation.children[0];
+	assert(child->type == AL_OPERAND);
+	GrB_Info info = GrB_transpose(res, GrB_NULL, GrB_NULL, child->operand.matrix, GrB_NULL);
+	assert(info == GrB_SUCCESS);
+	return res;
+}
+
+static GrB_Matrix _Eval_Add(const AlgebraicExpression *exp, GrB_Matrix res) {
+	assert(exp && AlgebraicExpression_ChildCount(exp) > 1);
+
+	GrB_Info info;
+	GrB_Index nrows;                // Number of rows of operand.
+	GrB_Index ncols;                // Number of columns of operand.
+	GrB_Matrix a = GrB_NULL;        // Left operand.
+	GrB_Matrix b = GrB_NULL;        // Right operand.
+	GrB_Matrix inter = GrB_NULL;    // Intermidate matrix.
+	GrB_Descriptor desc = GrB_NULL; // Descriptor used for transposing operands.
+	bool res_in_use = false;        // Can we use `res` for intermidate evaluation.
+
+	GrB_Descriptor_new(&desc);
+
+	// Get left and right operands.
+	AlgebraicExpression *left = CHILD_AT(exp, 0);
+	AlgebraicExpression *right = CHILD_AT(exp, 1);
+
+	/* If left operand is a matrix, simply get it.
+	 * Otherwise evaluate left hand side using `res` to store LHS value. */
+	if(left->type == AL_OPERAND) {
+		a = left->operand.matrix;
+	} else {
+		if(left->operation.op == AL_EXP_TRANSPOSE) {
+			a = left->operation.children[0]->operand.matrix;
+			GrB_Descriptor_set(desc, GrB_INP0, GrB_TRAN);
+		} else {
+			a = _AlgebraicExpression_Eval(left, res);
+			res_in_use = true;
+		}
+	}
+
+	/* If right operand is a matrix, simply get it.
+	 * Otherwise evaluate right hand side using `res` if free or create an additional matrix to store RHS value. */
+	if(right->type == AL_OPERAND) {
+		b = right->operand.matrix;
+	} else {
+		if(right->operation.op == AL_EXP_TRANSPOSE) {
+			b = right->operation.children[0]->operand.matrix;
+			GrB_Descriptor_set(desc, GrB_INP1, GrB_TRAN);
+		} else if(res_in_use) {
+			// `res` is in use, create an additional matrix.
+			GrB_Matrix_nrows(&nrows, a);
+			GrB_Matrix_ncols(&ncols, a);
+			info = GrB_Matrix_new(&inter, GrB_BOOL, nrows, ncols);
+			if(info != GrB_SUCCESS) {
+				fprintf(stderr, "%s", GrB_error());
+				assert(false);
+			}
+			b = _AlgebraicExpression_Eval(right, inter);
+		} else {
+			// `res` is not used just yet, use it for RHS evaluation.
+			b = _AlgebraicExpression_Eval(right, res);
+		}
+	}
+
+	// Perform addition.
+	if(GrB_eWiseAdd_Matrix_Semiring(res, GrB_NULL, GrB_NULL, GxB_LOR_LAND_BOOL, a, b,
```

Comment:
why semiring and not GrB_eWiseAdd_Matrix_BinaryOp with GrB_PLUS_BOOL

Reply:
It's all the same